### ISSUE TITLE
Mixing PHP with HTML

### ISSUE THREAD

**Issue by ichiriac** [wmhilton/beautify#4]
A common usecase in PHP is to mix HTML output with PHP code / templating.

The main problem here is that HTML is never formated by php-unparser, so resulting in ugly HTML output.

You could before formatting PHP code, pre-process it AS html, and then format it, resulting in a fully formater document.

Same thing could apply on javascript and css embed in HTML documents.

> NOTE : I've saw something already implemented but did not succeed to make it work

**Comment by billiegoose** [comment]
Yes indeed! I need a way to format the HTML inside those weird php blocks, like:
```php
<?php if (true) { ?>
  <b>some html</b>
<?php } ?>
```

The tough part is the HTML inside them can be fragments. They are not guaranteed to be complete nodes. E.g.

```php
<div><b>
<?php if (true) { ?>
  some html</b></div>
<?php } else { ?>
  other </b>html
<?php } ?>
```

And I just haven't wrapped my head around how to do it yet. I tried in the [tokenGetAll branch](https://github.com/wmhilton/beautify/commits/tokenGetAll) but that mostly broke things. In an *ideal* world, `php-unparser` would allow me to pass an HTML-formatter callback as an option. That would make it very simple.

**Comment by billiegoose** [comment]
Ideally it could even call the formatter with a desired indentation depth. Then the HTML formatter could preserve the correct indentation, making appropriate exceptions for `<pre> </pre>` tags and JavaScript ``  template string literals.

**Comment by ichiriac** [comment]
You could maybe handle each document type separately :

Here a complex sample :
```
<html lang="<?= $lang ?>">
  ...
  <?php 
if (something) :
?><head>
        <?php endif; ?>
  </head>
</html>
```

You can do it on 3 passes - 1st pass with html :
```
<html lang="<!--PHP:1-->">
  ...
  <!--PHP:2--><head>
        <!--PHP:3-->
  </head>
</html>

**** BECOMES :
<html lang="<!--PHP:1-->">
  ...
  <!--PHP:2-->
  <head>
    <!--PHP:3-->
  </head>
</html>
```

2nd pass with PHP :
```
H:1<?= $lang ?>H:2<?php 
if (something) :
?>H:3<?php endif; ?>H:4
**** SHOULD BECOME :
<?php
echo 'H:1';
echo $lang;
if ($something):
  echo 'H:3';
endif;
```

With the 1st result, and 2nd result you can extract parts, and create the 3rd pass :
```
<html lang="<?= $lang ?>">
  ...
  <?php if (something) : ?>
   <head>
    <?php endif; ?>
  </head>
</html>
```

Rules for the 3rd part are :

- you split the PHP file
- for each part you rebuild HTML, by respecting HTML indentation
- if the PHP splited part if multi-line, you add the HTML indentation on each PHP line

**Comment by billiegoose** [comment]
I'd really prefer if the php-unparser didn't turn HTML into echo statements. That's the root of the problem.
